text/temp/PALENICA/Larsson_Sadakane_Suffix_Array.tex

```diff
@@ -0,0 +1,113 @@
+
+\begin{lemma}
+Długość ścieżki od korzenia do liścia jest ograniczona z góry przez $2 \log{n} + 3$
+\begin{proof}
+Każdy węzeł `środkowy` na ścieżce odpowiada podwojeniu \textit{h-order} w którym znajduje się rozważana tablica. Takich węzłów może być więc $\log{n} + 1$ Z faktu, że używamy mediany po zejściu lewym lub prawym dzieckiem długość rozważanej tablicy zmniejsza się o połowę. Znów takich węzłów będzie maksymalnie $\log{n} + 1$
+\end{proof}
+\end{lemma}
+Na danym poziomie drzewa ilość wykonanej pracy będzie $O(n)$ bo to obliczenie odpowiada wykonaniu podziału na rozłączne podtablice. Stąd dostajemy złożoność $O(n\ \log{n})$ bo aktualizacja podtablic odpowiedzialnych za grupy również jest wykonywana w czasie liniowym od rozmiaru podtablicy.
+
+\section*{Usprawnienia}
+
+W algorytmie autorzy wprowadzają usprawnienia które nie wpływają negatywnie na czas wykonania. Opiszemy tutaj większość z nich (te które zostały wykorzystane w implementacji).
+
+\textbf{Eliminacja tablicy długości $L$}. Tablica długości dla grup nieposortowanych jest nam zbędna ze względu na to jak numerujemy grupy. Przypomnijmy, numer grupy $I[f \ldots g]$ to $g$, czyli $V[f-1] = g$ a w konsekwencji rozmiar grupy to $V[f-1] - g + 1$. Do spamiętania rozmiarów wspólnych grup posortowanych możemy użyć tablicy $I$. Wpólne grupy posortowane nie pojawią się w wywołaniu quicksorta a tylko tam korzystamy z $I$ w trakcie działania algorytmu. Jedynym problemem pozostaje odzyskanie porządku w $I$ na samym końcu algorytmu, ale tutaj wystarczy, że skorzystamy z tablicy $V$ trzymającej numery grup (indywidualne grupy posortowane mają różne numery), więc $I[V[i]] = i$ dla $0 \leq i < n+1$ odzyska tablicę sufiksową na końcu. Ostatecznie otrzymujemy następującą procedurę odzyskania długości grupy: jeśli $I[i] < 0$ to $I[i \ldots i - I[i] - 1]$ jest wspólną grupą posortowaną. W przeciwnym wypadku $I[i \ldots i + V[I[i]-1]]$ jest nieposortowaną grupą. 
+
+\textbf{Połączenie sortowania i aktualizowania tablic}. Zauważmy, że wykonanie połączenia dwóch następujących wspólnych grup posortowanych możemy wykonywać dopiero przy kroku 4 algorytmu gdy iterujemy się po kolejnych nieposortowanych grupach w $I$ przeskakując nad wspólnymi grupami posortowanymi. Aktualizacja tablic dla grup nieposortowanych w czasie wykonania sortowania musi być wykonana z uwagą na to żeby nie zmienić wartości w $V$ które będą wykorzystane jako klucze sortowania. Fakt wykorzystania quicksorta zapewnia nam, że ustalenie kolejności aktualizacji tablic jest proste. Niektóre elementy w pierwszej partycji będą wykorzystywać jako klucz numery grup elementów z środkowej partycji (bo bierzemy $V[I[currentElement] + h - 1]$ jako klucz). Tym samym aktualizację numerów grup z partycji środkowej wykonujemy po rekurencyjnym posortowaniu pierwszej partycji.
```

Comment:
"Niektóre elementy w pierwszej partycji będą wykorzystywać jako klucz numery grup elementów z środkowej partycji (bo bierzemy $V[I[currentElement] + h - 1]$ jako klucz)."
Well, since h is unrelated to the size of all three partitions after ternary sort, the indices might come from all three of them, so I don't see at all how that works.

Reply:
OK, changed to correct explanation. Thank you for the proof idea.